tempesta_fw/http_sess.c

```diff
@@ -0,0 +1,731 @@
+
+	spin_unlock(&hb->lock);
+
+	req->sess = sess;
+
+	return 0;
+}
+
+int __init
+tfw_http_sess_init(void)
+{
+	int ret;
+	u_char *ptr;
+
+	if ((ptr = kzalloc(STICKY_NAME_MAXLEN + 1, GFP_KERNEL)) == NULL) {
+		return -ENOMEM;
+	}
+	tfw_cfg_sticky.name.ptr = tfw_cfg_sticky.name_eq.ptr = ptr;
+	tfw_cfg_sticky.name.len = tfw_cfg_sticky.name_eq.len = 0;
+
+	tfw_sticky_shash = crypto_alloc_shash("hmac(sha1)", 0, 0);
+	if (IS_ERR(tfw_sticky_shash)) {
+		pr_err("shash allocation failed\n");
+		return  PTR_ERR(tfw_sticky_shash);
+	}
+
+	get_random_bytes(tfw_sticky_key, sizeof(tfw_sticky_key));
+	ret = crypto_shash_setkey(tfw_sticky_shash,
+				  (u8 *)tfw_sticky_key,
+				  sizeof(tfw_sticky_key));
+	if (ret) {
+		crypto_free_shash(tfw_sticky_shash);
+		return ret;
+	}
+
+	sess_cache = kmem_cache_create("tfw_sess_cache", sizeof(TfwHttpSess),
+				      0, 0, NULL);
+	if (!sess_cache) {
+		crypto_free_shash(tfw_sticky_shash);
+		return -ENOMEM;
+	}
+
+	return 0;
+}
+
+void
+tfw_http_sess_exit(void)
+{
+	int i;
+
+	for (i = 0; i < (1 << SESS_HASH_BITS); ++i) {
+		TfwHttpSess *s;
+		struct hlist_node *tmp;
+		SessHashBucket *hb = &sess_hash[i];
+
+		hlist_for_each_entry_safe(s, tmp, &hb->list, hentry) {
+			hash_del(&s->hentry);
+			kmem_cache_free(sess_cache, s);
+		}
+	}
+	kmem_cache_destroy(sess_cache);
+
+	kfree(tfw_cfg_sticky.name.ptr);
+	memset(&tfw_cfg_sticky, 0, sizeof(tfw_cfg_sticky));
+	crypto_free_shash(tfw_sticky_shash);
+}
+
+static int
+tfw_cfg_sess_start(void)
+{
+	tfw_cfg_sticky.enabled = !TFW_STR_EMPTY(&tfw_cfg_sticky.name);
+
+	return 0;
+}
+
+static void
+tfw_cfg_sess_stop(void)
+{
+	tfw_cfg_sticky.enabled = 0;
+}
+
+static int
+tfw_http_sticky_cfg(TfwCfgSpec *cs, TfwCfgEntry *ce)
+{
+	size_t i, len;
+	const char *val;
+
+	val = tfw_cfg_get_attr(ce, "name", STICKY_NAME_DEFAULT);
+	len = strlen(val);
+	if (len == 0 || len > STICKY_NAME_MAXLEN)
+		return -EINVAL;
+	memcpy(tfw_cfg_sticky.name.ptr, val, len);
+	tfw_cfg_sticky.name.len = len;
+	((char*)tfw_cfg_sticky.name_eq.ptr)[len] = '=';
+	tfw_cfg_sticky.name_eq.len = len + 1;
+
+	TFW_CFG_ENTRY_FOR_EACH_VAL(ce, i, val) {
+		 if (!strcasecmp(val, "enforce")) {
+			tfw_cfg_sticky.enforce = 1;
+			break;
+		}
+	}
+
+	return 0;
+}
+
+static int
+tfw_http_sticky_secret_cfg(TfwCfgSpec *cs, TfwCfgEntry *ce)
+{
+	int r, len = strlen(ce->vals[0]);
+
+	if (tfw_cfg_check_single_val(ce))
+		return -EINVAL;
+	if (len > STICKY_KEY_MAXLEN)
+		return -EINVAL;
+
+	memset(tfw_sticky_key, 0, STICKY_KEY_MAXLEN);
+	memcpy(tfw_sticky_key, ce->vals[0], len);
+	r = crypto_shash_setkey(tfw_sticky_shash, (u8 *)tfw_sticky_key,
+				STICKY_KEY_MAXLEN);
+	if (r) {
+		crypto_free_shash(tfw_sticky_shash);
+		return r;
+	}
+	return 0;
+}
+
+TfwCfgMod tfw_http_sess_cfg_mod = {
+	.name = "http_sticky",
+	.start = tfw_cfg_sess_start,
+	.stop = tfw_cfg_sess_stop,
+	.specs = (TfwCfgSpec[]) {
+		{
+			.name = "sticky",
+			.handler = tfw_http_sticky_cfg,
+			.allow_none = true,
+		},
+		{
+			.name = "sticky_secret",
```

Comment:
Is there the intention for the `sticky_secret` directive that it may somehow apply to all `sticky` directives? If not, then, perhaps, this should better be a part of the `sticky` directive:
`sticky name=abracadabra secret=blablabla enforce;`